crates/ruff_linter/src/rules/flake8_pyi/rules/unnecessary_type_union.rs

```diff
@@ -70,116 +75,132 @@ pub(crate) fn unnecessary_type_union<'a>(checker: &mut Checker, union: &'a Expr)
     let mut type_exprs: Vec<&Expr> = Vec::new();
     let mut other_exprs: Vec<&Expr> = Vec::new();
 
-    let mut collect_type_exprs = |expr: &'a Expr, _parent: &'a Expr| match expr {
-        Expr::Subscript(ast::ExprSubscript { slice, value, .. }) => {
-            if semantic.match_builtin_expr(value, "type") {
-                type_exprs.push(slice);
-            } else {
-                other_exprs.push(expr);
+    let mut has_pep604_union = false;
```

Comment:
Rather than using the same Union syntax as the top-level union, I check if _any_ PEP604-style union is encountered. If so, then we use that. (`a | b` or `typing.Union[a, b]`)